### PR TITLE
[XPU] Support Gemma-3 bidirectional image attention in XPUAttentionBackend

### PR DESCRIPTION
**Pull request by desmond-intel**
## Motivation

Gemma-3 multimodal uses **bidirectional attention among image tokens** during prefill (text stays causal). In SGLang this is realized by `Gemma3ForConditionalGeneration.prepare_attn_masks()` building a dense per-sequence mask — but that path is gated to `TritonAttnBackend` only (see the note in `gemma3_mm.py`: *"only the TritonAttnBackend supports bidirectional attention; other backends have not yet implemented this"*).

On the Intel XPU attention backend the mask was never applied, so image tokens ran plain **causal** attention, degrading vision-heavy tasks while leaving text unaffected.

On Intel Arc Pro B60 (`--attention-backend intel_xpu`, tp1), `docvqa_val` ANLS:

| backend | ANLS |
|---|---|
| intel_xpu (before) | 59.4 |
| triton (same HW) | 64.1 |
| H200 (ref) | 65.0 |
| **intel_xpu (this PR, sgl-kernel-xpu native)** | **67.48** |

Text-only `gsm8k_cot` is unchanged between intel_xpu and triton, consistent with the gap being image-specific.

## Modifications

This does the bidirectional masking **in the kernel** (via sgl-project/sgl-kernel-xpu#495), keeping the sglang path a normal flash-attn call — no attention fallback in the model runner.

- `models/gemma3_mm.py::prepare_attn_masks`: for `XPUAttentionBackend`, build a per extend-token block id (image index, `-1` for text) and store it on the backend (`is_xpu()`-guarded lazy import; the Triton path is unchanged). Full prefill only.
- `layers/attention/xpu_backend.py::forward_extend`: when present, run non-causal and pass `bidirectional_block_ids` to `flash_attn_with_kvcache`; the kernel re-imposes causal for block-less tokens and keeps same-block (image) tokens bidirectional. `window_size` is dropped for this path (short multimodal prefill: the sliding window is a no-op and would otherwise re-mask the block).
- `layers/layernorm.py::forward_xpu`: make the input contiguous before `gemma_rmsnorm` (it requires a contiguous last dim, which a sliced hidden state may not have) so Gemma-3 loads on XPU.

Depends on the companion kernel PR sgl-project/sgl-kernel-xpu#495 (`bidirectional_block_ids`). Run with `--chunked-prefill-size -1` (full prefill).

## Accuracy

`docvqa_val` (lmms-eval, ANLS), Intel Arc Pro B60, tp1, `--enable-multimodal --attention-backend intel_xpu --chunked-prefill-size -1`: **59.4 → 67.48** over the full 5349-sample val set, above the H200 reference (65.0). `gsm8k_cot` (text) unchanged. The kernel is unit-tested against an fp32 reference on the companion PR.

## Checklist

- [x] Lint: ruff / ruff-format / isort clean.
- [x] Bidirectional prefill is prefill-only; decode / CUDA-graph paths are untouched.

---
### CI States

Latest PR Test (Base): :x: [Run #35329506536](https://github.com/sgl-project/sglang/actions/runs/35329506536)
Latest PR Test (Extra): :x: [Run #35329506329](https://github.com/sgl-project/sglang/actions/runs/35329506329)
Latest PR Test (AMD ROCm 10): :x: [Run #35329506379](https://github.com/sgl-project/sglang/actions/runs/35329506379)

### CODE REVIEW

**Review by airMeng — COMMENTED** [review]
if a native fallback attention is needed, can you implement it in sgl-kernel-xpu and keep the interface in sglang unchanged after the performant solution ready?

File: python/sglang/srt/layers/attention/xpu_backend.py [hunk 1]
```diff
@@ -803,6 +846,85 @@ def forward_extend(
...
+            oi = F.scaled_dot_product_attention(
+                qi,
+                kf.unsqueeze(0),
+                vf.unsqueeze(0),
+                attn_mask=attn_mask,
+                scale=layer.scaling,
+            )  # [1, h_q, q_len, dv]
+            out[s:e] = oi.squeeze(0).transpose(0, 1)
+
+        return out.view(-1, h_q * dv)
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/xpu_backend.py]: bidirectional goes to triton on CUDA side.
at the currrent moment, you can also do so.

later you can provide support from sgl-kernel-xpu.

use torch fallback would be very slow.

File: python/sglang/srt/models/gemma3_mm.py [hunk 2]
```diff
@@ -221,8 +221,25 @@ def prepare_attn_masks(
...
+        if not is_triton:
+            try:
+                from sglang.srt.layers.attention.xpu_backend import (
+                    XPUAttentionBackend,
+                )
+
+                is_xpu = isinstance(backend, XPUAttentionBackend)
+            except Exception:
+                is_xpu = False
+
```
Comment (mingfeima) [inline python/sglang/srt/models/gemma3_mm.py]: catching every exception while importing `XPUAttentionBackend` silently restores causal image attention, which is the incorrect behavior this PR fixes.

Guard the import with `is_xpu()` and let unexpected XPU import errors fail loudly.

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/attention/xpu_backend.py (hunk 1 above)
Comment (desmond-intel) [reply python/sglang/srt/layers/attention/xpu_backend.py]: I've checked the perf of the torch fallback here,  it's actually not slow, and comes out faster than routing to triton on XPU:

gemma-3-4b-it `docvqa_val` (B60, tp1, limit-20, `--chunked-prefill-size -1`):

| | intel_xpu + SDPA | triton |
|---|---|---|
| prefill wall span | ~21s | ~56s |
| end-to-end (24 reqs) | ~28s | ~86s |
| prefill tok/s (median/mean) | 420 / 1475 | 64 / 874 |

Further, I'm working on `sgl-kernel-xpu` support for the proper solution.

File: python/sglang/srt/models/gemma3_mm.py (hunk 2 above)
Comment (desmond-intel) [reply python/sglang/srt/models/gemma3_mm.py]: Done, gated the import on `is_xpu()` and dropped the `try/except` so a real XPU import error surfaces instead of silently falling back to causal. (renamed the local flag to `is_xpu_backend` to avoid shadowing the helper.) e858b35132

**Comment by desmond-intel** [comment]
Updated this PR to the **native sgl-kernel-xpu** approach instead of the earlier torch/SDPA fallback, per the review (@mingfeima @airMeng). The sglang side is now thin, it builds a per extend-token image-block id and passes `mm_block_id` to `flash_attn_with_kvcache`; the bidirectional masking is done in the kernel.

Companion kernel PR: sgl-project/sgl-kernel-xpu#495 (adds the `mm_block_id` prefill mask). This PR depends on it.

- `prepare_attn_masks`: for `XPUAttentionBackend`, build the block ids (image index, `-1` for text); Triton path unchanged.
- `forward_extend`: run non-causal + pass `mm_block_id`; the kernel re-imposes causal for text and keeps image tokens bidirectional.
- `layernorm.forward_xpu`: make the input contiguous before `gemma_rmsnorm` (needs a contiguous last dim) so Gemma-3 loads.

Accuracy unchanged from the fallback: `docvqa_val` (full 5349) **59.4 -> 67.5** on Intel Arc Pro B60, above H200 (65.0). Kernel unit test added on the kernel PR.

The `is_xpu()` import guard from the previous round is retained.

File: python/sglang/srt/layers/layernorm.py [hunk 3]
```diff
@@ -1325,9 +1325,11 @@ def forward_xpu(self, x, residual: Optional[torch.Tensor] = None):
             gemma_fused_add_rmsnorm(x, residual, self.weight.data, self.eps)
             return x, residual
         # The XPU kernel flattens leading dims internally, so 2D/3D/4D inputs
-        # can all go through it directly without a Python-side reshape.
+        # can all go through it directly without a Python-side reshape. It does
+        # require a contiguous last dim, which a sliced/permuted hidden state may
+        # not have.
         elif residual is None and x.dim() in (2, 3, 4):
-            return gemma_rmsnorm(x, self.weight.data, self.eps)
+            return gemma_rmsnorm(x.contiguous(), self.weight.data, self.eps)
```
Comment (airMeng) [inline python/sglang/srt/layers/layernorm.py]: I think our implementation support in-contiguous output
https://github.com/sgl-project/sgl-kernel-xpu/blob/ecddde5da12abfc752b0aa9523e5ffb0bc5391d7/src/sycl/RMSNorm.cpp#L955

File: python/sglang/srt/models/gemma3_mm.py [hunk 4]
```diff
@@ -272,6 +273,40 @@ def prepare_attn_masks(
...
+            seq = torch.full((ext,), -1, dtype=torch.int32, device=device)
+            mm_inputs = forward_batch.mm_inputs[i]
+            if mm_inputs is not None:
+                for mm_item in mm_inputs.mm_items:
+                    if mm_item.is_image():
+                        for im_begin, im_end in mm_item.offsets:
+                            seq[im_begin : im_end + 1] = next_id
+                            next_id += 1
+            block_ids.append(seq)
+        return torch.cat(block_ids) if next_id > 0 else None
```
Comment (airMeng) [inline python/sglang/srt/models/gemma3_mm.py]: extract as a single funtion to sgl-kernel.

File: python/sglang/srt/layers/layernorm.py (hunk 3 above)
Comment (desmond-intel) [reply python/sglang/srt/layers/layernorm.py]: Checked the kernel: gemma_rmsnorm handles non-contiguous batch strides, but still needs a contiguous last dim (a transposed input trips CHECK_LAST_DIM_CONTIGUOUS), which Gemma-3 hits here. Tightened the guard to only make it contiguous when stride(-1) != 1, so the batch-strided cases the kernel already handles stay untouched. Can move it kernel-side (relax the last-dim check in gemma_rmsnorm) if you prefer. PTAL.

File: python/sglang/srt/models/gemma3_mm.py (hunk 4 above)
Comment (desmond-intel) [reply python/sglang/srt/models/gemma3_mm.py]: Extracted to sgl-kernel-xpu as build_bidirectional_block_ids(seq_lens, block_spans, device) in sgl-project/sgl-kernel-xpu#495. What stays here only reads the spans from ForwardBatch/MultimodalInputs and calls it (that part I feel can't move into the kernel package). PTAL.

File: python/sglang/srt/layers/layernorm.py [hunk 5]
```diff
@@ -1324,9 +1324,13 @@ def forward_xpu(self, x, residual: Optional[torch.Tensor] = None):
...
             return x, residual
-        # The XPU kernel flattens leading dims internally, so 2D/3D/4D inputs
-        # can all go through it directly without a Python-side reshape.
+        # The XPU kernel flattens leading dims internally, so 2D/3D/4D inputs go
+        # through directly without a Python-side reshape. It handles
+        # non-contiguous batch/row strides but requires a contiguous last dim,
+        # which a permuted hidden state may not have.
         elif residual is None and x.dim() in (2, 3, 4):
+            if x.stride(-1) != 1:
+                x = x.contiguous()
```
Comment (airMeng) [inline python/sglang/srt/layers/layernorm.py:1333]: which one is more effective: 1. reorder x. 2. fallback to native path if not contiguous, just like ```forward_cpu```

Comment (desmond-intel) [reply python/sglang/srt/layers/layernorm.py:1333]: Reorder (option 1) should win here. It copies only when stride(-1) != 1 (no-op otherwise), then stays on the fused kernel and keeps identical numerics for contiguous and non-contiguous inputs. The native fallback skips the copy but runs the slower multi-op path and can diverge numerically from the fused kernel for those tokens. Kept the reorder; can switch to a forward_native fallback if you prefer. PTAL.
